File: src/main/java/org/mitallast/queue/common/events/DefaultEventBus.java
package org.mitallast.queue.common.events;

import javaslang.collection.HashMultimap;
import javaslang.collection.Multimap;

import java.util.concurrent.Executor;
import java.util.concurrent.locks.ReentrantLock;
import java.util.function.Consumer;

public class DefaultEventBus implements EventBus {
    private final ReentrantLock lock = new ReentrantLock();
    private volatile Multimap<Class, Listener> consumers = HashMultimap.withSet().empty();

    public <Event> void subscribe(Class<Event> eventClass, Consumer<Event> consumer) {
        lock.lock();
        try {
            consumers = consumers.put(eventClass, new Listener<>(consumer));
        } finally {
            lock.unlock();
        }
    }

    @Override
    public <Event> void subscribe(Class<Event> eventClass, Consumer<Event> consumer, Executor executor) {
        lock.lock();
        try {
            consumers = consumers.put(eventClass, new Listener<>(consumer, executor));
        } finally {
            lock.unlock();
        }
    }

    @Override
    public <Event> void unsubscribe(Class<Event> eventClass, Consumer<Event> consumer) {
        lock.lock();
        try {
            consumers = consumers.remove(eventClass, new Listener<>(consumer));
        } finally {
            lock.unlock();
        }
    }

    @Override
    @SuppressWarnings("unchecked")
    public <Event> void trigger(Event event) {
        consumers.get(event.getClass()).forEach(consumers -> consumers.forEach(consumer -> consumer.accept(event)));
    }

    @Override
    @SuppressWarnings("unchecked")
    public <Event> void trigger(Class<Event> eventClass, Event event) {
        consumers.get(eventClass).forEach(consumers -> consumers.forEach(consumer -> consumer.accept(event)));
    }

    private static class Listener<Event> {
        private final Consumer<Event> consumer;
        private final Executor executor;

        public Listener(Consumer<Event> consumer) {
            this(consumer, null);
        }

        public Listener(Consumer<Event> consumer, Executor executor) {
            this.consumer = consumer;
            this.executor = executor;
        }

        public void accept(Event event) {
            if (executor == null) {
                consumer.accept(event);
            } else {
                executor.execute(() -> consumer.accept(event));
            }
        }

        @Override
        public boolean equals(Object o) {
            if (this == o) return true;
            if (o == null || getClass() != o.getClass()) return false;

            Listener<?> that = (Listener<?>) o;

            return consumer.equals(that.consumer);
        }

        @Override
        public int hashCode() {
            return consumer.hashCode();
        }
    }
}

File: src/main/java/org/mitallast/queue/common/events/DefaultEventBus.kt
package org.mitallast.queue.common.events

import javaslang.collection.HashMultimap
import javaslang.collection.Multimap
import java.util.concurrent.Executor
import java.util.concurrent.locks.ReentrantLock

@Suppress("UNCHECKED_CAST")
class DefaultEventBus : EventBus {
    private val lock = ReentrantLock()
    @Volatile private var consumers: Multimap<Class<*>, Listener<(Any) -> Unit>> = HashMultimap.withSet<Any>().empty()

    override fun <Event : Any> subscribe(eventClass: Class<Event>, consumer: (Event) -> Unit) {
        lock.lock()
        try {
            consumers = consumers.put(eventClass, Listener(consumer) as Listener<(Any) -> Unit>)
        } finally {
            lock.unlock()
        }
    }

    override fun <Event : Any> subscribe(eventClass: Class<Event>, consumer: (Event) -> Unit, executor: Executor) {
        lock.lock()
        try {
            consumers = consumers.put(eventClass, Listener(consumer, executor) as Listener<(Any) -> Unit>)
        } finally {
            lock.unlock()
        }
    }

    override fun <Event : Any> unsubscribe(eventClass: Class<Event>, consumer: (Event) -> Unit) {
        lock.lock()
        try {
            consumers = consumers.remove(eventClass, Listener(consumer) as Listener<(Any) -> Unit>)
        } finally {
            lock.unlock()
        }
    }

    override fun <Event : Any> trigger(event: Event) {
        consumers.get(event.javaClass).forEach { listeners ->
            listeners.forEach { listener ->
                (listener as Listener<Event>).accept(event)
            }
        }
    }

    private class Listener<in Event : Any> constructor(
        private val consumer: (Event) -> Unit,
        private val executor: Executor? = null) {

        fun accept(event: Event) {
            if (executor == null) {
                consumer.invoke(event)
            } else {
                executor.execute { consumer.invoke(event) }
            }
        }

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (javaClass != other?.javaClass) return false
            other as Listener<*>
            if (consumer != other.consumer) return false
            return true
        }

        override fun hashCode(): Int {
            return consumer.hashCode()
        }
    }
}
